## Neural Network using TensorFlow and Keras

___

### The MNIST dataset

The MNIST dataset consists of 70,000 28x28 pixel grayscale images of handwritten single digits from 0 to 9. The training set contains 60,000 images and the test set 10,000. For more information you can find the dataset [here](http://yann.lecun.com/exdb/mnist/).

<img src="MnistExamples.png" width=600 />


___

In this notebook we will use the following library:
- [tensorflow](https://www.tensorflow.org/install)

In [1]:
import tensorflow as tf

In [65]:
(train_x, train_y), (test_x, test_y) = tf.keras.datasets.mnist.load_data()

___
The color values of the pixels are 0 to 255. In the code cell below we will standardize these values to be 0 to 1.
___

In [66]:
train_x, test_x = train_x / 255.0, test_x / 255.0

In [78]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(100, activation='sigmoid'),
  tf.keras.layers.Dense(100, activation='sigmoid'),
  tf.keras.layers.Dense(100, activation='sigmoid'),
  tf.keras.layers.Dense(10)])

In [79]:
predictions = model(train_x[:1]).numpy()
predictions

array([[ 0.86674833, -0.8684917 ,  0.65524346,  0.22865975, -0.51648724,
        -0.12357749, -0.38304418, -0.7147247 , -0.23684785, -0.3661984 ]],
      dtype=float32)

In [80]:
tf.nn.softmax(predictions).numpy()

array([[0.23520608, 0.04148044, 0.19036765, 0.12425967, 0.05898166,
        0.08736862, 0.06740165, 0.04837528, 0.07801225, 0.0685467 ]],
      dtype=float32)

In [81]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [82]:
loss_fn(train_y[:1], predictions).numpy()

2.437619

In [83]:
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

In [84]:
model.fit(train_x, train_y, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.9703 - accuracy: 0.7170
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1910 - accuracy: 0.9444
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1212 - accuracy: 0.9639
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0921 - accuracy: 0.9723
Epoch 5/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0702 - accuracy: 0.9792


In [85]:
model.evaluate(test_x, test_y, verbose=2)

313/313 - 0s - loss: 0.0967 - accuracy: 0.9708


[0.09670677781105042, 0.97079998254776]

Running the model for 5 epochs we get an accuracy of approximately 98% on the training data. Evaluating the models performance on the test data we get an accuracy of 97%.